pkg/reconciler/buildrun/resources/taskrun.go

```diff
@@ -208,22 +215,23 @@ func GenerateTaskSpec(
 
 		generatedTaskSpec.Steps = append(generatedTaskSpec.Steps, step)
 
-		// Get volumeMounts added to Task's spec.Volumes
-		for _, volumeInBuildStrategy := range containerValue.VolumeMounts {
-			newVolume := true
-			for _, volumeInTask := range generatedTaskSpec.Volumes {
-				if volumeInTask.Name == volumeInBuildStrategy.Name {
-					newVolume = false
-				}
-			}
-			if newVolume {
-				generatedTaskSpec.Volumes = append(generatedTaskSpec.Volumes, corev1.Volume{
-					Name: volumeInBuildStrategy.Name,
-				})
+		for _, vm := range containerValue.VolumeMounts {
+			// here we should check that volume actually exists for this mount
+			// and in case it does not, exit early with an error
+			if _, ok := buildStrategyVolumesMap[vm.Name]; !ok {
+				return nil, fmt.Errorf("Volume for the Volume Mount %q is not found", vm.Name)
 			}
+			volumeMounts[vm.Name] = vm.ReadOnly
 		}
 	}
 
+	// Add volumes from the strategy to generated task spec
+	volumes, err := volumes.TaskSpecVolumes(volumeMounts, buildStrategyVolumes, build.Spec.Volumes, buildRun.Spec.Volumes)
+	if err != nil {
+		return nil, err
+	}
+	generatedTaskSpec.Volumes = volumes
```

Comment:
You must append here. `AmendTaskSpecWithSources` ends up in https://github.com/shipwright-io/build/blob/v0.9.0/pkg/reconciler/buildrun/resources/sources/utils.go#L30 when the source has a secret associated. The volume(s) added there, you would erase here, I think.

Reply:
My bad. Thank you for noticing that!